frontend/src/App.js

```diff
@@ -1,94 +1,107 @@
-        updatedOnlineOnMount.current = true;
-        updateLastOnlineMutation.mutate();
-      }
-      
-      const interval = setInterval(() => {
-        updateLastOnlineMutation.mutate();
-      }, 60000);
-
-      return () => {
-        clearInterval(interval);
-      };
-    }
-  }, [currentUser, updateLastOnlineMutation]);
-
-  return (
-    <BrowserRouter>
-      {currentUser?.initialData ? ( <LoadingPage /> ) : ( 
-        <Routes>
-          {homeRoute}
-          {adminRoutes}
-          {userRoutes}
-          {courseRoutes}
-          <Route path="*" element={<NotFoundPage />} />
-        </Routes>
-      )}
-    </BrowserRouter>
-  );
-}
-
-export default App;
+import { BrowserRouter, Routes, Route } from "react-router-dom";
+import { useEffect, useRef } from "react"
+
+import { useBackendMutation } from "main/utils/useBackend";
+import HomePage from "main/pages/HomePage";
+
+import LoadingPage from "main/pages/LoadingPage";
+import LoginPage from "main/pages/LoginPage";
+import ProfilePage from "main/pages/ProfilePage";
+import CoursesEditPage from "main/pages/CoursesEditPage";
+
+import AdminUsersPage from "main/pages/AdminUsersPage";
+import AdminJobsPage from "main/pages/AdminJobsPage";
+import SchoolIndexPage from "main/pages/SchoolIndexPage";
+
+import CoursesCreatePage from "main/pages/CoursesCreatePage";
+import CourseIndexPage from "main/pages/CourseIndexPage";
+
+import StaffCreatePage from "main/pages/StaffCreatePage";
+import StaffEditPage from "main/pages/StaffEditPage";
+import StaffIndexPage from "main/pages/StaffIndexPage";
+
+import { hasRole, useCurrentUser } from "main/utils/currentUser";
+import NotFoundPage from "main/pages/NotFoundPage";
+
+function App() {
+  const { data: currentUser } = useCurrentUser();
+
+  const adminRoutes = hasRole(currentUser, "ROLE_ADMIN") ? (
+    <>
+      <Route path="/admin/schools" element={<SchoolIndexPage />} />
+      <Route path="/admin/users" element={<AdminUsersPage />} />
+      <Route path="/admin/jobs" element={<AdminJobsPage />} />
+    </>
+  ) : null;
+
+  const userRoutes = hasRole(currentUser, "ROLE_USER") ? (
+    <>
+      <Route path="/profile" element={<ProfilePage />} />
+      <Route path="/courses" element={<CourseIndexPage />} />
+    </>
+  ) : null;
+
+  const courseRoutes = (hasRole(currentUser, "ROLE_ADMIN") || hasRole(currentUser, "ROLE_INSTRUCTOR")) ? (
+    <>
+      <Route path="/courses/create" element={<CoursesCreatePage />} />
+      <Route path="/courses" element={<CourseIndexPage />} />
+      <Route path="/courses/edit/:id" element={<CoursesEditPage />} />
+    </>
+  ) : null;
+
+  const staffRoutes = hasRole(currentUser, "ROLE_ADMIN") || hasRole(currentUser, "ROLE_INSTRUCTOR") ? (
+    <>
+      <Route path="/courses/:courseId/staff/create" element={<StaffCreatePage />} />
+      <Route path="/courses/:courseId/staff" element={<StaffIndexPage />} />
+      <Route path="/courses/:courseId/staff/edit/:id" element={<StaffEditPage />} />

```

Comment:
```suggestion
```

There's no use case for edit for a staff member.  We add and delete.  There's nothing to edit, since the three fields are id (not editable), courseId (not editable), and githubId (not editable).